google/cloud/bigtable/completion_queue.cc

```diff
@@ -19,12 +19,83 @@ namespace google {
 namespace cloud {
 namespace bigtable {
 inline namespace BIGTABLE_CLIENT_NS {
+namespace {
+/**
+ * Wrap a gRPC timer into an `AsyncOperation`.
+ *
+ * Applications (or more likely, other components in the client library) will
+ * associate timers with a completion queue. gRPC timers require applications to
+ * create a unique `grpc::Alarm` object for each timer, and then to associate
+ * them with the completion queue using a `void*` tag.
+ *
+ * This class collaborates with our wrapper for `CompletionQueue` to associate
+ * a `future<AsyncTimerResult>` for each timer. This class takes care of
+ * allocating the `grpc::Alarm`, creating a unique `void*` associated with the
+ * timer, and satisfying the future when the timer expires.
+ *
+ * Note that this class is an implementation detail, hidden from the application
+ * developers.
+ */
+class AsyncTimerFuture : public internal::AsyncGrpcOperation {
+ public:
+  explicit AsyncTimerFuture(std::unique_ptr<grpc::Alarm> alarm)
+      : alarm_(std::move(alarm)) {}
+
+  future<std::chrono::system_clock::time_point> GetFuture() {
+    return promise_.get_future();
+  }
+
+  void Set(grpc::CompletionQueue& cq,
+           std::chrono::system_clock::time_point deadline, void* tag) {
+    std::unique_lock<std::mutex> lk(mu_);
+    deadline_ = deadline;
+    if (alarm_) {
+      alarm_->Set(&cq, deadline, tag);
+    }
+  }
+
+  void Cancel() override {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (alarm_) {
+      alarm_->Cancel();
+    }
+  }
+
+ private:
+  bool Notify(CompletionQueue&, bool) override {
+    std::unique_lock<std::mutex> lk(mu_);
+    alarm_.reset();
```

Comment:
Prefer assignment (although in this case just of `nullptr`) over `.reset()` to clarify the disposition of the `deleter`.

Reply:
Done.